### PR TITLE
Codechange: Use functions to access colour gradients and replace magic numbers.

### PR DESCRIPTION
**Pull request by PeterN**
## Motivation / Problem

Mapping from `Colours` to pixel colour uses direct access to the `_colour_gradient` array, along with an additional index for the shade.

There is limited bounds checking for this access, and due to an off-by-one in TTD, the shades range from 1 to 7 instead of 0 to 7.

## Description

Instead, implement and use functions `GetColourGradient()` and `SetColourGradient()`. These ensure all access is within bounds.

A new enum is added to allow the shade to be specified by name, and the off-by-one is removed (technically a fix but it did not affect anything so not marked as a fix) which allows access to the light(er)est shade.

## Checklist for review

Some things are not automated, and forgotten often. This list is a reminder for the reviewers.
* The bug fix is important enough to be backported? (label: 'backport requested')
* This PR touches english.txt or translations? Check the [guidelines](https://github.com/OpenTTD/OpenTTD/blob/master/docs/eints.md)
* This PR affects the save game format? (label 'savegame upgrade')
* This PR affects the GS/AI API? (label 'needs review: Script API')
    * ai_changelog.hpp, game_changelog.hpp need updating.
    * The compatibility wrappers (compat_*.nut) need updating.
* This PR affects the NewGRF API? (label 'needs review: NewGRF')
    * newgrf_debug_data.h may need updating.
    * [PR must be added to API tracker](https://wiki.openttd.org/en/Development/NewGRF/Specification%20Status)

### CODE REVIEW

File: src/misc_gui.cpp [hunk 1]
```diff
@@ -794,7 +794,7 @@ void QueryString::DrawEditBox(const Window *w, WidgetID wid) const
 
 	DrawFrameRect(cr, wi->colour, wi->IsLowered() ? FR_LOWERED : FR_NONE);
 	DrawSpriteIgnorePadding(rtl ? SPR_IMG_DELETE_RIGHT : SPR_IMG_DELETE_LEFT, PAL_NONE, cr, SA_CENTER);
-	if (this->text.bytes == 1) GfxFillRect(cr.Shrink(WidgetDimensions::scaled.bevel), _colour_gradient[wi->colour & 0xF][2], FILLRECT_CHECKER);
```
Comment (2TallTyler) [inline src/misc_gui.cpp:797]: What was this `colour & 0xF` doing? (here and elsewhere)

File: src/company_cmd.cpp [hunk 2]
```diff
@@ -146,8 +146,8 @@ void SetLocalCompany(CompanyID new_company)
  */
 TextColour GetDrawStringCompanyColour(CompanyID company)
 {
-	if (!Company::IsValidID(company)) return (TextColour)_colour_gradient[COLOUR_WHITE][4] | TC_IS_PALETTE_COLOUR;
-	return (TextColour)_colour_gradient[_company_colours[company]][4] | TC_IS_PALETTE_COLOUR;
+	if (!Company::IsValidID(company)) return (TextColour)GetColourGradient(COLOUR_WHITE, SHADE_NORMAL) | TC_IS_PALETTE_COLOUR;
```
Comment (2TallTyler) [inline src/company_cmd.cpp:149]: Can these be static_casts?

File: src/misc_gui.cpp (hunk 1 above)
Comment (PeterN) [reply src/misc_gui.cpp:797]: It was making sure the value of `wi->colour` is within bounds (not always done) which is now done by `GetColourGradient()`

Comment (2TallTyler) [reply src/misc_gui.cpp:797]: Ah, thanks.

**Review by rubidium42 — APPROVED** [review]

**Pull request merged**